### PR TITLE
Support assume role credential and instance profile.

### PR DESCRIPTION
**Pull request by toyama0919**
I respond assume role credeential and instance profile.

Usage is as follow.(example)

## assume role

```
[default]
role_arn = arn:aws:iam::549338637798:role/assume-role-test
source_profile = default
region = ap-northeast-1
```

## instance profile

```
[default]
credential_source=Ec2InstanceMetadata
```

### CODE REVIEW

**Comment by sonots** [comment]
ref. https://docs.aws.amazon.com/cli/latest/topic/config-vars.html#using-aws-iam-roles

```
# In ~/.aws/credentials:
[development]
aws_access_key_id=foo
aws_secret_access_key=bar

# In ~/.aws/config
[profile crossaccount]
role_arn=arn:aws:iam:...
source_profile=development
```

```
# In ~/.aws/config
[profile crossaccount]
role_arn=arn:aws:iam:...
credential_source=Ec2InstanceMetadata
```

File: lib/ec2/host/ec2_client.rb [hunk 1]
```diff
@@ -37,11 +38,31 @@ def ec2
...
         if Config.aws_access_key_id and Config.aws_secret_access_key
           Aws::Credentials.new(Config.aws_access_key_id, Config.aws_secret_access_key)
-        else
-          Aws::SharedCredentials.new(profile_name: Config.aws_profile, path: Config.aws_credential_file)
+        else # use profile
+          shared_credentials = Aws::SharedCredentials.new(
+            profile_name: Config.aws_profile,
+            path: Config.aws_credential_file
+          )
+          config = Aws.shared_config.instance_variable_get(:@parsed_config)[Config.aws_profile]
```
Comment (sonots) [inline lib/ec2/host/ec2_client.rb]: I dislike to use `instance_variable_get`.
I investigated about AWS SDK, but it looked AWS::SharedConfig class itself is also private https://github.com/aws/aws-sdk-ruby/blob/7c9584bc6473100df9aec9333ab491ad4faeeca8/gems/aws-sdk-core/lib/aws-sdk-core/shared_config.rb#L3-L4.

Referring codenize-tools https://github.com/codenize-tools/piculet/pull/28/files, it was using https://github.com/a2ikm/aws_config gem. Do you think it is good idea to use this gem (depend on this gem)?

Comment (toyama0919) [reply lib/ec2/host/ec2_client.rb]: > https://github.com/codenize-tools/piculet/pull/28/files

This is good example!
i'll try it.

**Comment by toyama0919** [comment]
Done.

* Use aws_config gem.
* remove inifile gem.

(piculet not support InstanceProfileCredentials..

File: lib/ec2/host/ec2_client.rb [hunk 2]
```diff
@@ -38,7 +39,21 @@ def credentials
         if Config.aws_access_key_id and Config.aws_secret_access_key
           Aws::Credentials.new(Config.aws_access_key_id, Config.aws_secret_access_key)
         else
-          Aws::SharedCredentials.new(profile_name: Config.aws_profile, path: Config.aws_credential_file)
+          config = Config.aws_config
```
Comment (sonots) [inline lib/ec2/host/ec2_client.rb]: Prefer `aws_config = Config.aws_config` because `config` in EC2::Host should represent EC2::Host::Config.

Comment (toyama0919) [reply lib/ec2/host/ec2_client.rb]: OK. Thanks.

**Comment by sonots** [comment]
lgtm

**Comment by sonots** [comment]
v1.0.1 is released

**Comment by sonots-zozo** [comment]
Hmm, I noticed that a combination like:

* ~/.aws/config with region
* instance profile

is not supported well. I anyway feature requested to https://github.com/a2ikm/aws_config/issues/9, but I feel we should fix the behavior of ec2-host.

**Comment by sonots-zozo** [comment]
Because the instance profile config is written in `~/.aws/config` as:

```
[default]
region=ap-northeast-1
credential_source=Ec2InstanceMetadata
```

`~/.aws/credentials` does not exist usually

**Comment by sonots-zozo** [comment]
I may prefer not to have even `credential_source=Ec2InstanceMetadata`. Just fallback to instance profile if all authentication failed since aws cli works so. Is it possible?

File: lib/ec2/host/ec2_client.rb (hunk 1 above)
Comment (sonots) [reply lib/ec2/host/ec2_client.rb]: Now, I noticed that it was possible to obtain this as `Config.aws_config[:role_arn]`

Comment (sonots) [reply lib/ec2/host/ec2_client.rb]: I will remove introducing `aws_config` gem.

File: lib/ec2/host/config.rb [hunk 3]
```diff
@@ -65,13 +65,10 @@ def self.aws_config_file
...
       def self.aws_config
         return @aws_config if @aws_config
-        if File.readable?(aws_config_file)
-          ini = IniFile.load(aws_config_file).to_h
-          if aws_profile == 'default'
-            @aws_config = ini['default']
-          else
-            @aws_config = ini["profile #{aws_profile}"]
-          end
+        if File.readable?(aws_config_file) && File.readable?(aws_credential_file)
```
Comment (toyama0919) [inline lib/ec2/host/config.rb:68]: ここを 

```
if File.readable?(aws_config_file)
```

にしたらどうなるんだろう。
configの方から読んでくれそうな気が。

**Comment by toyama0919** [comment]
>Just fallback to instance profile if all authentication failed. Is it possible?

naruhodo.

File: lib/ec2/host/config.rb (hunk 3 above)
Comment (sonots) [reply lib/ec2/host/config.rb:68]: 動かなかったょ

**Comment by sonots** [comment]
Fixed via https://github.com/sonots/ec2-host/pull/17

File: lib/ec2/host/config.rb (hunk 3 above)
Comment (toyama0919) [reply lib/ec2/host/config.rb:68]: そうでしたか、、あざす

**Pull request merged**